sql: TestSecondaryIndexWithOldStoringEncoding failed

sql.TestSecondaryIndexWithOldStoringEncoding [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21052175?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21052175?buildTab=artifacts#/) on master @ [0d609bc851df2d9332468d62a8408d2e5c3ac21e](https://github.com/cockroachdb/cockroach/commits/0d609bc851df2d9332468d62a8408d2e5c3ac21e):


```
          | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
          | 	pkg/sql/colexecerror/error.go:162
          | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
          | 	pkg/sql/colflow/stats.go:130
          | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter
          | 	pkg/sql/colflow/flow_coordinator.go:242
          | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
          | 	pkg/sql/colexecerror/error.go:162
          | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next
          | 	pkg/sql/colflow/flow_coordinator.go:246
          | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run
          | 	pkg/sql/colflow/flow_coordinator.go:278
          | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
          | 	pkg/sql/colflow/vectorized_flow.go:317
          | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
          | 	pkg/sql/distsql_running.go:1087
          | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
          | 	pkg/sql/distsql_running.go:2353
          | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3
          | 	pkg/sql/distsql_running.go:2052
          | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
          | 	pkg/sql/distsql_running.go:2055
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
          | 	pkg/sql/conn_executor_exec.go:3452
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
          | 	pkg/sql/conn_executor_exec.go:2951
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
          | 	pkg/sql/conn_executor_exec.go:1088
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func2
          | 	pkg/sql/conn_executor_exec.go:178
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling.func1
          | 	pkg/sql/conn_executor_exec.go:4502
          | github.com/cockroachdb/cockroach/pkg/util/pprofutil.Do
          | 	pkg/util/pprofutil/labels.go:38
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
          | 	pkg/sql/conn_executor_exec.go:4499
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
          | 	pkg/sql/conn_executor_exec.go:177
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
          | 	pkg/sql/conn_executor.go:2413
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
          | 	pkg/sql/conn_executor.go:2418
          | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
          | 	pkg/sql/conn_executor.go:2326
        Wraps: (2) insufficient bytes to decode varint value: "\x04"
        Error types: (1) *withstack.withStack (2) *errutil.leafError
--- FAIL: TestSecondaryIndexWithOldStoringEncoding/data_scan#01 (0.02s)
=== RUN   TestSecondaryIndexWithOldStoringEncoding/index_scan
    schema_changer_test.go:3748: expected "fetched: /t/i/1/11/2 -> <undecoded>" but read "fetched: /t/i/1/11/0 -> <undecoded>"
--- FAIL: TestSecondaryIndexWithOldStoringEncoding/index_scan (0.01s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSecondaryIndexWithOldStoringEncoding.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-59277

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -3693,24 +3693,32 @@ CREATE TABLE d.t (
 	tableDesc := desctestutils.TestingGetMutableExistingTableDescriptor(
 		kvDB, server.Codec(), "d", "t")
 	// Verify that this descriptor uses the new STORING encoding. Overwrite it
-	// with one that uses the old encoding.
-	for _, index := range tableDesc.PublicNonPrimaryIndexes() {
-		if index.NumKeySuffixColumns() != 1 {
-			t.Fatalf("KeySuffixColumnIDs not set properly: %s", tableDesc)
+	// with one that uses the old encoding. Use TestingDescsTxn so that the
+	// descriptor version is properly incremented and the lease manager waits
+	// for all old leases to drain before proceeding.
+	if err := sqltestutils.TestingDescsTxn(context.Background(), server, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
+		tbl, err := col.MutableByID(txn.KV()).Table(ctx, tableDesc.GetID())
+		if err != nil {
+			return err
 		}
-		if index.NumSecondaryStoredColumns() != 1 {
-			t.Fatalf("StoreColumnIDs not set properly: %s", tableDesc)
+		for _, index := range tbl.PublicNonPrimaryIndexes() {
+			if index.NumKeySuffixColumns() != 1 {
+				t.Fatalf("KeySuffixColumnIDs not set properly: %s", tbl)
+			}
+			if index.NumSecondaryStoredColumns() != 1 {
+				t.Fatalf("StoreColumnIDs not set properly: %s", tbl)
+			}
+			newIndexDesc := index.IndexDescDeepCopy()
+			newIndexDesc.KeySuffixColumnIDs = append(newIndexDesc.KeySuffixColumnIDs, newIndexDesc.StoreColumnIDs...)
+			newIndexDesc.StoreColumnIDs = nil
+			tbl.SetPublicNonPrimaryIndex(index.Ordinal(), newIndexDesc)
 		}
-		newIndexDesc := index.IndexDescDeepCopy()
-		newIndexDesc.KeySuffixColumnIDs = append(newIndexDesc.KeySuffixColumnIDs, newIndexDesc.StoreColumnIDs...)
-		newIndexDesc.StoreColumnIDs = nil
-		tableDesc.SetPublicNonPrimaryIndex(index.Ordinal(), newIndexDesc)
-	}
-	if err := kvDB.Put(
-		context.Background(),
-		catalogkeys.MakeDescMetadataKey(server.Codec(), tableDesc.GetID()),
-		tableDesc.DescriptorProto(),
-	); err != nil {
+		ba := txn.KV().NewBatch()
+		if err := col.WriteDescToBatch(ctx, false /* kvTrace */, tbl, ba); err != nil {
+			return err
+		}
+		return txn.KV().Run(ctx, ba)
+	}); err != nil {
 		t.Fatal(err)
 	}
 	if _, err := sqlDB.Exec(`INSERT INTO d.t VALUES (11, 1, 2);`); err != nil {

--- a/pkg/testutils/sqlutils/inspect.go
+++ b/pkg/testutils/sqlutils/inspect.go
@@ -9,7 +9,6 @@ import (
 	gosql "database/sql"
 	"fmt"
 	"strings"
-	"time"
 
 	"github.com/cockroachdb/errors"
 )
@@ -22,7 +21,7 @@ type InspectResult struct {
 	Table      string
 	PrimaryKey string
 	JobID      int64
-	Aost       time.Time
+	Aost       string
 	Details    string
 }
 

